anaconda_anon_usage/heartbeat.py

```diff
@@ -9,56 +9,64 @@
 import argparse
 import os
 import sys
+import time
 from threading import Thread
 from urllib.parse import urljoin
 
-from conda.base.context import context
+from conda.base.context import Context, context, locate_prefix_by_name
 from conda.gateways.connection.session import get_session
 from conda.models.channel import Channel
 
 from . import utils
 
 VERBOSE = False
 STANDALONE = False
-DRY_RUN = os.environ.get("ANACONDA_HEARTBEAT_DRY_RUN")
 
 CLD_REPO = "https://repo.anaconda.cloud/"
 ORG_REPO = "https://conda.anaconda.org/"
 COM_REPO = "https://repo.anaconda.com/pkgs/"
 REPOS = (CLD_REPO, COM_REPO, ORG_REPO)
 HEARTBEAT_PATH = "noarch/activate-0.0.0-0.conda"
 
+# How long to attempt the connection. When a connection to our
+# repository is blocked or slow, a long timeout would lead to
+# a slow activation and a poor user experience. This is a total
+# timeout value, inclusive of all retries.
+TIMEOUT = 0.75  # seconds
+ATTEMPTS = 3
 
-def _print(msg, *args, standalone=False, error=False):
+
+def _print(msg, *args, error=False):
     global VERBOSE
     global STANDALONE
     if not (VERBOSE or utils.DEBUG or error):
         return
-    if standalone and not STANDALONE:
-        return
     # It is very important that these messages are printed to stderr
     # when called from within the activate script. Otherwise they
     # will insert themselves into the activation command set
     ofile = sys.stdout if STANDALONE and not (error or utils.DEBUG) else sys.stderr
     print(msg % args, file=ofile)
 
 
-def _ping(session, url, wait):
+def _ping(session, url, timeout):
     try:
-        response = session.head(url, proxies=session.proxies)
-        _print("Status code (expect 404): %s", response.status_code)
+        # A short timeout is necessary here so that the activation
+        # is not unduly delayed by a blocked internet connection
+        start_time = time.perf_counter()
+        response = session.head(url, proxies=session.proxies, timeout=timeout)
+        delta = time.perf_counter() - start_time
+        _print(
+            "Success after %.3fs; code (expect 404): %d", delta, response.status_code
+        )
     except Exception as exc:
         if type(exc).__name__ != "ConnectionError":
-            _print("Heartbeat error: %s", exc, error=True)
+            _print("Unexpected heartbeat error: %s", exc, error=True)
+        elif "timeout=" in str(exc):
+            delta = time.perf_counter() - start_time
+            _print("NO heartbeat sent after %.3fs.", delta)
 
 
-def attempt_heartbeat(channel=None, path=None, wait=False):
-    global DRY_RUN
-    line = "------------------------"
```

Comment:
moved the print statements, which are only needed when calling from the command line, to the `main()` function where they belong